bounded-collections/src/bounded_btree_map.rs

```diff
@@ -40,10 +41,15 @@ where
 	S: Get<u32>,
 {
 	fn decode<I: codec::Input>(input: &mut I) -> Result<Self, codec::Error> {
-		let inner = BTreeMap::<K, V>::decode(input)?;
-		if inner.len() > S::get() as usize {
+		// Same as the underlying implementation for `Decode` on `BTreeMap`, except we fail early if
+		// the len is too big.
+		let len: u32 = <Compact<u32>>::decode(input)?.into();
+		if len > S::get() {
 			return Err("BoundedBTreeMap exceeds its limit".into())
 		}
+		input.descend_ref()?;
+		let inner = Result::from_iter((0..len).map(|_| Decode::decode(input)))?;
```

Comment:
No idea if rust handles this in the most efficient way, otherwise we should maybe pre-allocate as optimization.

Reply:
I mean this is now discusable with the bound, but for the normal `BTreeMap` we don't have done this as someone could send you a fake encoded `BTreeMap` that pretends to be `u32::Max` in size and then maybe leading to some OOM.

Reply:
Interesting, would have to compile with optimizations and see. At any rate, I don't want to deviate too much from the standard `Decode` for the unbounded types, at least in this PR.

Reply:
Should be fine to merge as is.

Reply:
Is this just another way of calling `collect`?

```suggestion
		let inner = (0..len).map(|_| Decode::decode(input)).collect::<Result<Vec<_>, codec::Error>>()?;
```

Reply:
Yes, I think it is. http://doc.rust-lang.org/1.66.1/std/result/enum.Result.html#method.from_iter

Currently this impl is the same as the unbounded version, apart from the length check. I think it should be kept the same unless there's a good reason not to.